dynamicPermissions.ps1

```diff
@@ -0,0 +1,464 @@
+        [void]$headers.Add('Authorization', "$($AccessToken.token_type) $($AccessToken.access_token)")
+        [void]$headers.Add('Accept', 'application/json')
+        [void]$headers.Add('Content-Type', 'application/json')
+        Write-Output $headers
+    }
+    catch {
+        throw $_
+    }
+}
+#endregion functions
+
+try {
+    # Create authorization headers
+    try {
+        $headers = New-AuthorizationHeaders -ClientId $clientId -ClientSecret $clientSecret
+    }
+    catch {
+        $ex = $PSItem
+        $errorMessage = Get-ErrorMessage -ErrorObject $ex
+
+        Write-Verbose "Error at Line [$($ex.InvocationInfo.ScriptLineNumber)]: $($ex.InvocationInfo.Line). Error: $($errorMessage.VerboseErrorMessage)"
+
+        $outputContext.AuditLogs.Add([PSCustomObject]@{
+                # Action  = "" # Optional
+                Message = "Error creating authorization headers. Error Message: $($errorMessage.AuditErrorMessage)"
+                IsError = $true
+            })
+
+        # Throw terminal error
+        throw
+    }
+
+    # Get groups
+    try {
+        Write-Verbose "Querying groups"
+        $splatWebRequest = @{
+            Uri             = "$baseUrl/scim/groups"
+            Headers         = $headers
+            Method          = 'GET'
+            ContentType     = "application/json;charset=utf-8"
+            UseBasicParsing = $true
+        }
+        $groups = $null
+        $groups = (Invoke-RestMethod @splatWebRequest -Verbose:$false).resources
+
+        # Group on externalId to check if group exists (as externalId has to be unique for a group)
+        $groupsGrouped = $groups | Group-Object externalId -AsHashTable -AsString
+
+        Write-Information "Successfully queried groups. Result count: $(($groups | Measure-Object).Count)"
+    }
+    catch {
+        $ex = $PSItem
+        $errorMessage = Get-ErrorMessage -ErrorObject $ex
+
+        Write-Verbose "Error at Line [$($ex.InvocationInfo.ScriptLineNumber)]: $($ex.InvocationInfo.Line). Error: $($errorMessage.VerboseErrorMessage)"
+        Write-Verbose "URI: $($splatWebRequest.Uri)"
+
+        $outputContext.AuditLogs.Add([PSCustomObject]@{
+                # Action  = "" # Optional
+                Message = "Error querying groups. Error Message: $($errorMessage.AuditErrorMessage)"
+                IsError = $true
+            })
+
+        # Throw terminal error
+        throw
+    }
+
+    $desiredPermissions = @{ }
+    if (-Not($actionContext.Operation -eq "revoke")) {
+        # Example: Contract Based Logic:
+        foreach ($contract in $personContext.Person.Contracts) {
+            Write-Verbose "Contract in condition: $($contract.Context.InConditions)"
+            if ($contract.Context.InConditions -OR ($actionContext.DryRun -eq $True)) {
+                try {
+                    # Example: department_<department externalId>
+                    $groupExternalId = "department_" + $contract.Department.ExternalId
```

Comment:
better to use the dynamic construction like:
$correlationProperty = "Name"
$correlationValue = "department_" + $contract.Department.ExternalId

Reply:
That isn't possible, as we use a hashtable in which the groups are indexed on the external id.
One possibility is to add the correlationProperty to the top of the script, to make it easier, but that is still in the grouping.
I did change the $groupExternalId to a more generic name "$correlationValue"